compiler/src/dotty/tools/dotc/config/CommandLineParser.scala

```diff
@@ -1,68 +1,96 @@
- *  done correctly and portably.
+/** A simple enough command line parser.
  */
-object CommandLineParser {
-  // splits a string into a quoted prefix and the rest of the string,
-  // taking escaping into account (using \)
-  // `"abc"def` will match as `DoubleQuoted(abc, def)`
-  private class QuotedExtractor(quote: Char) {
-    def unapply(in: String): Option[(String, String)] = {
-      val del = quote.toString
-      if (in startsWith del) {
-        var escaped = false
-        val (quoted, next) = (in substring 1) span {
-          case `quote` if !escaped => false
-          case '\\'    if !escaped => escaped = true; true
-          case _                   => escaped = false; true
-        }
-        // the only way to get out of the above loop is with an empty next or !escaped
-        // require(next.isEmpty || !escaped)
-        if (next startsWith del) Some((quoted, next substring 1))
-        else None
-      }
-      else None
-    }
-  }
-  private object DoubleQuoted extends QuotedExtractor('"')
-  private object SingleQuoted extends QuotedExtractor('\'')
-  @sharable private val Word = """(\S+)(.*)""".r
-
-  // parse `in` for an argument, return it and the remainder of the input (or an error message)
-  // (argument may be in single/double quotes, taking escaping into account, quotes are stripped)
-  private def argument(in: String): Either[String, (String, String)] = in match {
-    case DoubleQuoted(arg, rest) => Right((arg, rest))
-    case SingleQuoted(arg, rest) => Right((arg, rest))
-    case Word(arg, rest)         => Right((arg, rest))
-    case _                       => Left(s"Illegal argument: $in")
-  }
-
-  // parse a list of whitespace-separated arguments (ignoring whitespace in quoted arguments)
-  @tailrec private def commandLine(in: String, accum: List[String] = Nil): Either[String, (List[String], String)] = {
-    val trimmed = in.trim
-    if (trimmed.isEmpty) Right((accum.reverse, ""))
-    else argument(trimmed) match {
-      case Right((arg, next)) =>
-        (next span Character.isWhitespace) match {
-          case("", rest) if rest.nonEmpty => Left("Arguments should be separated by whitespace.") // TODO: can this happen?
-          case(ws, rest)                  => commandLine(rest, arg :: accum)
-        }
-      case Left(msg) => Left(msg)
-    }
-  }
+object CommandLineParser:
+  inline private val DQ  = '"'
+  inline private val SQ  = '\''
+  inline private val EOF = -1
 
-  class ParseException(msg: String) extends RuntimeException(msg)
+  /** Split the line into tokens separated by whitespace or quotes.
+   *
+   *  Invoke `errorFn` with message on bad quote.
+   */
+  def tokenize(line: String, errorFn: String => Unit): List[String] =
+
+    var accum: List[String] = Nil
+
+    var pos   = 0
+    var start = 0
+    val qpos  = new ArrayBuffer[Int](16)    // positions of paired quotes
+
+    inline def cur    = if done then EOF else line.charAt(pos): Int
```

Comment:
Why is that type ascription needed? Could it be moved to be the result type of cur instead?
```suggestion
    inline def cur: Int = if done then EOF else line.charAt(pos)
```

Reply:
`charAt` is a char. The benefits of this style choice are to point out the problematic expression; it looks more cool; in result type position, it's like having eyeglasses wrapped in tape where they broke. That's why implicits, the nerdiest feature, requires it.

Reply:
ok, I guess that's fine.

Reply:
It's a bit quirky, and it was so long ago, I had to reconstruct my thinking. On the up side, with all the talk about command line parsing support for main entry point, probably this code will be replaced anyway. It was mostly aimed at Process building.

Reply:
Also thanks, this was to get my feet wet with the repo and process, low benefit and low risk.

Reply:
Thanks to you for porting this! And sorry for the long delay before the review, it slipped my mind.